### PR TITLE
Merge 'features/AnnotatedTypes' into 'features/NullableReferenceTypes'

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@alekseyts please review
cc @dotnet/roslyn-compiler

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs [hunk 1]
```diff
@@ -406,20 +402,11 @@ private ImmutableArray<TypeSymbolWithAnnotations> GetResults()
 
             // For now erase all notion about nullability of reference types as the algorithm is not taking it into account yet.
             var builder = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance(_fixedResults.Length);
-            bool erase = (((CSharpParseOptions)_binderOpt?.Compilation.SyntaxTrees.FirstOrDefault()?.Options)?.IsFeatureEnabled(MessageID.IDS_FeatureStaticNullChecking) == true);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: What is the advantage of passing this as a parameter instead?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: I see, the _binderOpt got removed.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 2]
```diff
@@ -427,15 +427,15 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
...
-            if (_returnInferenceCache.TryGetValue(paramInfo, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType)
+            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType)
             {
                 lambdaSymbol = returnInferenceLambda.Symbol;
-                var lambdaRetType = lambdaSymbol.ReturnType.TypeSymbol;
-                if ((object)LambdaSymbol.InferenceFailureReturnType != lambdaRetType &&
-                    lambdaRetType == returnType && lambdaSymbol.RefKind == refKind)
+                var lambdaReturnType = lambdaSymbol.ReturnType;
+                if ((object)LambdaSymbol.InferenceFailureReturnType != lambdaReturnType &&
+                    lambdaReturnType.Equals(returnType, TypeCompareKind.CompareNullableModifiersForReferenceTypes) && lambdaSymbol.RefKind == refKind)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: I don't remember witch comparison is stronger, I believe we should use strongest flavor here.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 3]
```diff
@@ -584,22 +584,40 @@ public override bool Equals(object obj)
...
+                return true;
             }
 
             public override int GetHashCode()
             {
-                return Hash.Combine(TaskLikeReturnTypeOpt?.GetHashCode() ?? 0, Hash.CombineValues(this.ParameterTypes));
+                var value = TaskLikeReturnTypeOpt?.GetHashCode() ?? 0;
+                foreach (var type in ParameterTypes)
+                {
+                    value = Hash.Combine(type.TypeSymbol, value);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Should we pass value as the first argument?

File: src/Compilers/CSharp/Portable/Errors/LasyUseSiteDiagnosticsInfoForNullableType.cs [hunk 4]
```diff
@@ -6,11 +6,11 @@
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
-    internal sealed class LasyUseSiteDiagnosticsInfoForNullableType : LasyDiagnosticsInfo
+    internal sealed class LazyUseSiteDiagnosticsInfoForNullableType : LazyDiagnosticsInfo
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Errors/LasyUseSiteDiagnosticsInfoForNullableType.cs]: File name should probably changed as well.

File: src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs [hunk 5]
```diff
@@ -147,6 +147,9 @@ internal TypeSymbolWithAnnotations SubstituteTypeWithTupleUnification(TypeSymbol
                 var possiblyTuple = TupleTypeSymbol.TransformToTupleIfCompatible(type);
                 if ((object)type != possiblyTuple)
                 {
+                    // PROTOTYPE(NullableReferenceTypes): This ignores the particular TypeSymbolWithAnnotations
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs]: We also might want to avoid resolving type eagerly several lines above ```var type = result.TypeSymbol```. Please add a comment for that as well.

File: src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs [hunk 6]
```diff
@@ -313,31 +313,31 @@ public override TResult Accept<TResult>(CSharpSymbolVisitor<TResult> visitor)
...
             if ((object)other == null || !other.HasSameShapeAs(this) || 
-                !other.ElementType.Equals(ElementType, options))
+                !other.ElementType.Equals(ElementType, comparison))
             {
                 return false;
             }
 
             // Make sure bounds are the same.
-            if ((options & TypeSymbolEqualityOptions.IgnoreArraySizesAndLowerBounds) == 0 && !this.HasSameSizesAndLowerBoundsAs(other))
+            if ((comparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0 && !this.HasSameSizesAndLowerBoundsAs(other))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs]: Are the options equivalent? It feels like IgnoreArraySizesAndLowerBounds might not ignore modifiers.

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs [hunk 7]
```diff
@@ -492,7 +484,7 @@ public int GetHashCode(Symbol member)
                 }
 
                 if (_considerReturnType && member.GetMemberArity() == 0 && 
-                    (_typeSymbolEqualityOptions & TypeSymbolEqualityOptions.IgnoreCustomModifiers) != 0) // If it is generic, then type argument might be in return type.
+                    (_typeComparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) != 0) // If it is generic, then type argument might be in return type.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: It looks like the options might be not equivalent.

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs [hunk 8]
```diff
@@ -540,7 +532,7 @@ private static bool HaveSameReturnTypes(Symbol member1, TypeMap typeMap1, Symbol
 
             if (isVoid1)
             {
-                if ((typeSymbolEqualityOptions & TypeSymbolEqualityOptions.IgnoreCustomModifiers) != 0 || 
+                if ((typeComparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) != 0 ||
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: It looks like the options might be not equivalent.

File: src/Compilers/CSharp/Portable/Symbols/Source/CustomModifierUtils.cs [hunk 9]
```diff
@@ -95,13 +95,11 @@ internal static TypeSymbol CopyTypeCustomModifiers(TypeSymbol sourceType, TypeSy
             bool transformResult = resultType.ApplyNullableTransforms(flagsBuilder.ToImmutableAndFree(), ref position, out resultType);
             Debug.Assert(transformResult && position == length);
 
-            Debug.Assert(resultType.Equals(sourceType, TypeSymbolEqualityOptions.IgnoreDynamic)); // Same custom modifiers as source type.
+            Debug.Assert(resultType.Equals(sourceType, TypeCompareKind.IgnoreDynamicAndTupleNames)); // Same custom modifiers as source type.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/CustomModifierUtils.cs]: It looks like the options might be not equivalent.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs [hunk 10]
```diff
@@ -335,10 +335,10 @@ private TypeSymbolWithAnnotations GetTypeSymbol()
                 // If we got a valid result that was not void then use the inferred type
                 // else create an error type.
                 if ((object)inferredType != null &&
-                    inferredType.SpecialType != SpecialType.System_Void)
+                    inferredType.TypeSymbol.SpecialType != SpecialType.System_Void)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: This change looks unnecessary, TypeSymbolWithAnnotations has SpecialType property?

File: src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs [hunk 11]
```diff
@@ -279,27 +279,27 @@ public TypeSymbolWithAnnotations AsNullableReferenceOrValueType(CSharpCompilatio
...
 
-            if ((object)other == null || !TypeSymbolEquals(other, options))
+            if ((object)other == null || !TypeSymbolEquals(other, comparison))
             {
                 return false;
             }
 
             // Make sure custom modifiers are the same.
-            if ((options & TypeSymbolEqualityOptions.IgnoreCustomModifiers) == 0 && !this.CustomModifiers.SequenceEqual(other.CustomModifiers))
+            if ((comparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0 && !this.CustomModifiers.SequenceEqual(other.CustomModifiers))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs]: It looks like the options might be not equivalent.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 12]
```diff
@@ -41,7 +38,7 @@ public NullableAttribute(bool[] transformFlags)
     [AttributeUsage(AttributeTargets.Module, AllowMultiple = true)]
     class NullableOptOutForAssemblyAttribute : Attribute
     {
-        /// <param name=""assemblyName"">An assembly name - a simple name plus its PublicKey, if any.""/></param>
+                    /// <param name=""assemblyName"">An assembly name - a simple name plus its PublicKey, if any.""/></param>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: This looks like an unexpected formatting change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 13]
```diff
@@ -177,15 +172,15 @@ class B : A
...
 ";
-            var compilation = CreateCompilationWithMscorlib(new[] { source, attributesDefinitions }, options: TestOptions.ReleaseDll, parseOptions: TestOptions.Regular.WithFeature("staticNullChecking", "true"));
+            var compilation = CreateCompilationWithMscorlib(new[] { source, attributesDefinitions }, options: TestOptions.ReleaseDll, parseOptions: TestOptions.Regular.WithNullCheckingFeature());
 
-            //var a = compilation.GetTypeByMetadataName("A");
-            //var aFoo = a.GetMember<MethodSymbol>("Foo");
-            //Assert.Equal("T? A.Foo<T>()", aFoo.ToTestDisplayString());
+                        //var a = compilation.GetTypeByMetadataName("A");
+                        //var aFoo = a.GetMember<MethodSymbol>("Foo");
+                        //Assert.Equal("T? A.Foo<T>()", aFoo.ToTestDisplayString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: This looks like an unexpected formatting change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 14]
```diff
@@ -177,15 +172,15 @@ class B : A
...
 
-            //var b = compilation.GetTypeByMetadataName("B");
-            //var bFoo = b.GetMember<MethodSymbol>("Foo");
-            //Assert.Equal("T? A.Foo<T>()", bFoo.OverriddenMethod.ToTestDisplayString());
+                        //var b = compilation.GetTypeByMetadataName("B");
+                        //var bFoo = b.GetMember<MethodSymbol>("Foo");
+                        //Assert.Equal("T? A.Foo<T>()", bFoo.OverriddenMethod.ToTestDisplayString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: This looks like an unexpected formatting change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 15]
```diff
@@ -500,9 +495,9 @@ public override void M1<T>(T? x)
     }
 } 
 ";
-            var compilation = CreateCompilationWithMscorlib(new[] { source, attributesDefinitions }, options: TestOptions.ReleaseDll, parseOptions: TestOptions.Regular.WithFeature("staticNullChecking", "true"));
+            var compilation = CreateCompilationWithMscorlib(new[] { source, attributesDefinitions }, options: TestOptions.ReleaseDll, parseOptions: TestOptions.Regular.WithNullCheckingFeature());
 
-            // TODO: The overriding is ambigous. We simply matched the first candidate. Should this be an error?
+                        // TODO: The overriding is ambigous. We simply matched the first candidate. Should this be an error?
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: This looks like an unexpected formatting change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 16]
```diff
@@ -1392,45 +1387,45 @@ class B1 : IA
 
 interface IB
 {
-    //event System.Action<string> E1; 
-    //event System.Action<string>? E2; 
+                //event System.Action<string> E1; 
+                //event System.Action<string>? E2;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: This looks like an unexpected formatting change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 17]
```diff
@@ -1392,45 +1387,45 @@ class B1 : IA
...
+                //event System.Action<string>? E2; 
     event System.Action<string?>? E3; 
 }
 
 class B2 : IB
 {
-    //event System.Action<string?> IB.E1; // 2
-    //event System.Action<string> IB.E2; // 2
+                //event System.Action<string?> IB.E1; // 2
+                //event System.Action<string> IB.E2; // 2
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: This looks like an unexpected formatting change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 18]
```diff
@@ -2169,7 +2164,7 @@ public void Overriding_25()
   .method public hidebysig specialname rtspecialname 
           instance void  .ctor() cil managed
   {
-    // Code size       8 (0x8)
+                // Code size       8 (0x8)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: This looks like an unexpected formatting change. There a several similar changes in this file.

File: src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs [hunk 19]
```diff
@@ -2389,8 +2389,8 @@ .maxstack  0
             var t1 = test1.Parameters[0].Type;
             var t2 = test2.Parameters[0].Type;
 
-            Assert.False(t1.Equals(t2, TypeSymbolEqualityOptions.None));
-            Assert.False(t2.Equals(t1, TypeSymbolEqualityOptions.None));
+            Assert.False(t1.Equals(t2, TypeCompareKind.AllIgnoreOptions));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs]: It looks like the options are not equivalent.

File: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs [hunk 20]
```diff
@@ -10,13 +10,16 @@ namespace Microsoft.CodeAnalysis
...
         IgnoreCustomModifiersAndArraySizesAndLowerBounds = 1,
         IgnoreDynamic = 2,
         IgnoreTupleNames = 4,
         IgnoreDynamicAndTupleNames = IgnoreDynamic | IgnoreTupleNames,
+        CompareNullableModifiersForReferenceTypes = 8,
+        UnknownNullableModifierMatchesAny = 16, // Has no impact without CompareNullableModifiersForReferenceTypes
+        AllAspects = CompareNullableModifiersForReferenceTypes, // PROTOTYPE(NullableReferenceTypes): Remove if not used.
         AllIgnoreOptions = IgnoreCustomModifiersAndArraySizesAndLowerBounds | IgnoreDynamic | IgnoreTupleNames,
-        AllIgnoreOptionsForVB = IgnoreCustomModifiersAndArraySizesAndLowerBounds | IgnoreTupleNames
+        AllIgnoreOptionsForVB = AllIgnoreOptions // PROTOTYPE(NullableReferenceTypes): Can be removed and replaced with AllIgnoreOptions.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs]: It is not clear what is the motivation for this change, we should not be including bits that VB doesn't know about.

File: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs [hunk 21]
```diff
@@ -10,13 +10,16 @@ namespace Microsoft.CodeAnalysis
     [Flags]
     internal enum TypeCompareKind
     {
-        ConsiderEverything = 0,
+        ConsiderEverything = 0, // PROTOTYPE(NullableReferenceTypes): Does this include comparing nullable modifiers? If not, rename.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs]: It feels like the comment should include the answer to the question: "Does this include comparing nullable modifiers?" Rather than the question.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: I believe `CompareNullableModifiersForReferenceTypes` is the strongest comparison.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Arguments are `T newKeyPart, int currentKey`. The second argument is expected to be a hash code.

File: src/Compilers/CSharp/Portable/Errors/LasyUseSiteDiagnosticsInfoForNullableType.cs (hunk 4 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Errors/LasyUseSiteDiagnosticsInfoForNullableType.cs]: Renamed file.

File: src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs (hunk 5 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs]: Added comment.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs (hunk 10 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs]: Reverted.

File: src/Compilers/CSharp/Portable/Symbols/Source/CustomModifierUtils.cs (hunk 9 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/CustomModifierUtils.cs]: We want to avoid comparing dynamic-ness and tuple names in this case.

File: src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs (hunk 6 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs]: This check is used to determine whether to compare sizes and bounds, not custom modifiers.

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: `IgnoreCustomModifiersAndArraySizesAndLowerBounds` is a single bit in master.

Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: Changed to `(_typeComparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0`.

File: src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs (hunk 19 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs]: Fixed.

File: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs (hunk 20 above)
Comment (cston) [inline src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs]: Reverted. But is there anything preventing a caller in VB from using `AllIgnoreOptions` instead of `AllIgnoreOptionsForVB` even with the extra enum value?

File: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs (hunk 21 above)
Comment (cston) [inline src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs]: Updated comment.

File: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs (hunk 20 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs]: That is a good point, perhaps we should remove ```AllIgnoreOptions``` member. CC @jcouv

Comment (jcouv) [inline src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs]: Just to clarify, you're suggesting renaming `AllIgnoreOptions` to `AllIgnoreOptionsForCSharp`?
That seems fine.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs]: > Just to clarify, you're suggesting renaming AllIgnoreOptions to AllIgnoreOptionsForCSharp?

That is correct. For some reason I thought we already had this member for C#.
